### ISSUE TITLE
Identify any work in IKS to leverage the recent zone/locality support

### ISSUE THREAD

**Issue by linsun** [istio/istio#11845]
We need to identify how we can leverage the recent zone support to ensure the zone info from IKS nodes are passed correctly to the envoy sidecars when users use istio within IKS.  Is there anything we need to populate when bootstrap when users run multiZone IKS clusters, similarly like https://github.com/istio/istio/pull/11607/files

cc @ericvn

**Comment by ymesika** [comment]
cc @vadimeisenbergibm

**Comment by ericvn** [comment]
It seems I can get the region and zone from `kubectl get nodes -o yaml` from the returned: 
```
      failure-domain.beta.kubernetes.io/region: us-south
      failure-domain.beta.kubernetes.io/zone: dal10
```
These same constants were added in #13927.

Will work on a code change to make the call and get the data.

**Comment by ericvn** [comment]
Some simple code to look at proxy version for IKS and getting zone and region:
```
func main() {
	// creates the in-cluster config
	config, err := rest.InClusterConfig()
	if err != nil {
		panic(err.Error())
	}
	// creates the clientset
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		panic(err.Error())
	}

        // Get Nodes, but will only look at one
	nodes, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{})
	if err != nil {
		panic(err.Error())
	}
	
	if len(nodes.Items) == 0 {
		fmt.Printf("No Nodes found\n")
		return
	}

	for _, node := range nodes.Items {
		nodeKubeProxyVersion := node.Status.NodeInfo.KubeProxyVersion
		onIKS := strings.HasSuffix(nodeKubeProxyVersion, "+IKS")
		if onIKS {
			fmt.Println("OnIKS: true")
		} else {
			fmt.Println("OnIKS: false")
			return
		}
		onIKS_2 := false
		var region, zone string
		for k, v := range node.Labels {
			if strings.HasPrefix(k, "ibm-cloud.kubernetes.io") {
				onIKS_2 = true
			}
			if k == "failure-domain.beta.kubernetes.io/region" {
				region = v
			}
			if k == "failure-domain.beta.kubernetes.io/zone" {
				zone = v
			}
		}
		if onIKS_2 != onIKS {
			fmt.Println("ERROR: - IKS from label is not equal to KubeProxyVersion")
			return
		}
		if region == "" || zone == "" {
			fmt.Printf("ERROR: - Unable to find region (%s) or zone (%s) on IKS \n", region, zone)
			return
		}
		fmt.Printf("IKS: region (%s) and zone (%s)\n", region, zone)
                return
	}
}

```

**Comment by ericvn** [comment]
And this will need a ClusterRole and ClusterRoleBinding (if we don't already) for RBAC to access the node list from the API.

**Comment by howardjohn** [comment]
You don't need the locality in the bootstrap, Pilot will already read from the node labels. https://github.com/istio/istio/blob/ed792dda8df438667ec80be319a296e0aca6584e/pilot/pkg/serviceregistry/kube/controller.go#L317

**Comment by linsun** [comment]
That is great - thank you for chiming in @howardjohn!  @ericvn could you validate this works properly with IKS, e.g. reading the node labels and configure it properly in our env?

**Comment by ericvn** [comment]
Moving back to milestone 1.3 as I will validate this is working properly now.

**Comment by ericvn** [comment]
Testing shows that the localityLbSetting's do work on IKS.

I did this work on the _master_ (soon to be the (_1.3_)) branch as it has localityLbSetting set to enabled:true by default. Note that no locality actions will be done without the specification of a outlierDetection in a DestinationRule (also need a service specified for the caller).

I deployed a multi-zone cluster on IKS using the `Shared control plane (multi-network)` instructions. Let's call WDC04 region _a_ and WDC06 region _b_. At the conclusion of the install we showed that calling curl on helloworld from sleep(_a_) resulted in calls to both helloworld(_a_) and helloworld(_b_), and calling curl on helloworld from sleep(_b_) resulted in calls to both helloworld(_a_) and helloworld(_b_).

I then created a DestinationRule:
```
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: helloworld-outlier-detection
spec:
  host: helloworld.sample.svc.cluster.local
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
    outlierDetection:
      consecutiveErrors: 7
      interval: 30s
      baseEjectionTime: 30s
```
and applied it to the control plane cluster: `kubectl apply --context=$CTX_CLUSTER1 -f helloworld-destination-rule.yaml -n sample`. I did not need to restart anything.

This enables the locality loadbalancer and calls should now stay within their regions.

The same curl commands run during the sample exercise do stay within their own region, ie. calling curl on helloworld from sleep(_b_) resulted in calls to helloworld(_b_) and calling curl on helloworld from sleep(_a_) resulted in calls to helloworld(_a_).

To test failover, I first tried scaling down the helloworld(b) application and curls from sleep(_b_), but they timed out. I was told that their is an issue with scaling down applications removing endpoints so this would NOT make helloworld(_b_) seem _unhealthy_, resulting in the timeouts. I scaled the replicas back to 1 and things worked as before. I then removed the application helloworld(_b_) and the curls in sleep(_b_) failed over to helloworld(_a_).

It's not easy to set the distribution in the localityLBsetting's. It might be with the new operator, and I will investigate that later. This precluded me from trying to distribute the calls from WDC04 to WDC06 and vice versa which should have caused the curls from sleep(_a_) to go to helloworld(_b_) and vice versa.

I do plan to write a blog extending the current sample (or an update in the doc) to talk about showing this behavior as the current docs are hard to interpret (and the current 1.3 doc still uses the old experimental flag).
